raspberryPI3/ros2_ws/src/detection/test/test_integration_obstacles_detection.py

```diff
@@ -0,0 +1,124 @@
+
+    # Callback function of the subscriber
+    def detected_message(self, msg):
+        self.received_data.append(msg.data)
+
+    # Beginning of the test
+    # /!\This method needs to start by "test_" to be interpreted by colcon test/!\
+    def test_node_behavior(self):
+        '''
+        We precise here the different inputs of the test
+        We can make several one in one time to avoid redundancy
+        We test several cases : above, below and equal to the detection limit
+        Try to make user relevant test (like acceptance tests in the backlog)
+        (no need to limit test like in an unit test)
+        Ex :
+        Given the vehicule is moving forward
+        When an obstacle is detected
+        Then the obstacle_detection_node needs to publish true messages
+        '''
+        test_cases = [
+            {"front_left": 40, "front_right": 40, "expected_output": True},
+            {"front_left": 80, "front_right": 80, "expected_output": False},
+            {"front_left": 50, "front_right": 50, "expected_output": True},
+            {"front_left": 60, "front_right": 60, "expected_output": False}
+        ]
+        # front_left, front_right and expected_output are not usable variables here
+        for case in test_cases:
+            with self.subTest(front_left=case["front_left"],
+                              front_right=case["front_right"],
+                              expected_output=case["expected_output"]
+                              ):
+                # You have to clear received_data to avoid interferences between tests, they are not independant
+                self.received_data.clear()
+
+                expected = case["expected_output"]
+                input_msg = Ultrasonic()
+                input_msg.front_left = case["front_left"]
+                input_msg.front_right = case["front_right"]
+                input_msg.front_center = 0
+                input_msg.rear_center = 0
+                input_msg.rear_left = 0
+                input_msg.rear_right = 0
+
+                # We wait for the node obstacles_detection to start (especially on my computer xD)
+                # And to let the time for obstacles_detection node to stop publishing True/false data on the topic
+                # So that u avoid interferences between test (especially relevant on my computer tbh)
+                time.sleep(5)  # secondes
+
+                # Then we publish the ultrasonic datas on /us_data
+                self.publisher.publish(input_msg)
+
+                # We wait for an answer of the obstacles_detection node
+                timeout = 15.0  # secondes
+                start_time = self.test_node.get_clock().now().nanoseconds / 1e9
+                while not self.received_data:
+                    rclpy.spin_once(self.test_node, timeout_sec=0.1)
+                    if (self.test_node.get_clock().now().nanoseconds / 1e9 - start_time) > timeout:
+                        break
```

Comment:
Could you make a method for that?  (it should look something like this)
```py
# def
def wait_for(self, data, timeout):
        while not self.received_data:
            rclpy.spin_once(self.test_node, timeout_sec=0.1)
             if (self.test_node.get_clock().now().nanoseconds / 1e9 - start_time) > timeout:
                    break

...
# use
 self.wait_for(self.received_data, 15.0)
````

Later we could improve some methods so that the tests are quicker and it won't change the tests itself. Actually it would ever be to make another class and users will inherit from it (like `unittest.TestCase`). So we change one method and it will be applied to all integration tests made by other developers